File: Week 07/id_358/counting-bits.js
/**
 * @param {number} num
 * @return {number[]}
 * 奇数：1 个数为前一个数的1 的个数+1
 * 偶数：1个数等于下一个乘以2的数的 1 的个数
 * DP方程
 * a[0] = 0
 * if n % 2 === 1  a[n] = a[n-1] + 1
 * if n % 2 === 0  a[n] = a[n/2]
 */
 var countBits = function(num) {
    let a = new Array(num+1)
    a[0] = 0;
    for(let i = 1; i <= num; i++) {
        if(i%2 === 1) {
            a[i] = a[i-1] + 1;
        } else if(i % 2 === 0) {
            a[i] = a[i/2]
        }
    }
    return a
};

File: Week 07/id_303/CountingBits.swift
//
//  CountingBits.swift

import Foundation


func countBits(_ num: Int) -> [Int] {
    guard num > 0 else {
        return [0]
    }
    
    var result = [0]
    var i = 0
    var total = 1
    
    for j in 1...num {
        result.append(result[i] + 1)
        i += 1
        if i == total {
            i = 0
            total = j + 1
        }
    }
    return result
}
